File: src/test/java/com/ealanta/productapp/web/LabelledProductControllerTest.java
package com.ealanta.productapp.web;

import com.ealanta.productapp.label.LabelType;
import com.ealanta.productapp.label.LabelledProduct;
import com.ealanta.productapp.label.LabelledProductSource;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;

import java.io.IOException;
import java.util.Arrays;

import static com.ealanta.productapp.web.LabelledProductController.LABEL_TYPE;
import static com.ealanta.productapp.web.LabelledProductController.PATH;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(SpringRunner.class)
@WebMvcTest(LabelledProductController.class)
public class LabelledProductControllerTest {

    private static String JSON = MediaType.APPLICATION_JSON_UTF8_VALUE;

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    LabelledProductSource labelledProductSource;

    @Captor
    ArgumentCaptor<LabelType> argLabelType;

    LabelledProduct lprod1;
    LabelledProduct lprod2;
    LabelledProduct lprod3;

    @Before
    public void setup() throws IOException {
        lprod1 = new LabelledProduct();
        lprod2 = new LabelledProduct();
        lprod3 = new LabelledProduct();

        when(labelledProductSource.getProducts(argLabelType.capture())).thenReturn(Arrays.asList(lprod1, lprod2, lprod3));
    }

    private ResultActions checkBasicStuff(ResultActions ra) {
        try {
            return ra.andDo(print())
                           .andExpect(status().isOk())
                           .andExpect(content().contentType(JSON));
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    private ResultActions doGet() {
        try {
            return mockMvc.perform(get(PATH));
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    private ResultActions doGetWithLabelType(String labelType) {
        try {
            return mockMvc.perform(get(PATH).param(LABEL_TYPE, labelType));
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    @Test
    public void testGetNoLabelTypeQueryString() {
        checkBasicStuff(doGet()).andReturn();
        Assert.assertEquals(LabelType.ShowWasNow, argLabelType.getValue());
    }

    private void checkGetWithLableType(LabelType type) {
        checkBasicStuff(doGetWithLabelType(type.name()));
        Assert.assertEquals(type, argLabelType.getValue());
        verify(labelledProductSource, times(1)).getProducts(argLabelType.getValue());
        verifyNoMoreInteractions(labelledProductSource);
    }

    @Test
    public void testWithInvalidLabelType() throws Exception {
        doGetWithLabelType("BOB").andDo(print())
                .andExpect(status().isUnprocessableEntity())
                .andExpect(content().contentTypeCompatibleWith(MediaType.TEXT_PLAIN))
                .andExpect(content().string("The parameter labelType must have a value among : ShowWasNow, ShowWasThenNow, ShowPercDiscount"));
    }

    @Test
    public void testGetWithWasNow() throws Exception {
        checkGetWithLableType(LabelType.ShowWasNow);
    }

    @Test
    public void testGetWithWasThenNow() throws Exception {
        checkGetWithLableType(LabelType.ShowWasThenNow);
    }

    @Test
    public void testGetWithPercDiscount() throws Exception {
        checkGetWithLableType(LabelType.ShowPercDiscount);
    }


}

File: src/src/test/kotlin/com/ealanta/productapp/web/LabelledProductControllerTest.kt
package com.ealanta.productapp.web

import com.ealanta.productapp.domain.LabelType
import com.ealanta.productapp.domain.LabelledProduct
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.ResultActions
import org.hamcrest.Matchers.`is`

import kotlin.properties.Delegates

import com.ealanta.productapp.web.LabelledProductController.Companion.LABEL_TYPE
import com.ealanta.productapp.web.LabelledProductController.Companion.PATH
import com.nhaarman.mockito_kotlin.*
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import kotlin.test.assertEquals

@RunWith(SpringRunner::class)
@WebMvcTest(LabelledProductController::class)
class LabelledProductControllerTest {

    companion object {
        private const val JSON = MediaType.APPLICATION_JSON_UTF8_VALUE
    }

    @Autowired
    lateinit private var mockMvc: MockMvc

    @MockBean
    lateinit private var mLabelledProductsSource : (LabelType) -> List<LabelledProduct>

    private var argCaptor: KArgumentCaptor<LabelType> by Delegates.notNull()

    private val labelledProduct1 = LabelledProduct().apply {
        priceLabel="PRICE LABEL 1"
        productId="PRODUCT ID 1"
    }

    private val labelledProduct2 = LabelledProduct().apply {
        priceLabel="PRICE LABEL 2"
        productId="PRODUCT ID 2"
    }

    private val labelledProduct3 = LabelledProduct().apply {
        priceLabel="PRICE LABEL 3"
        productId="PRODUCT ID 3"
    }

    @Before
    fun setup() {
        labelledProduct3
        argCaptor = argumentCaptor()
        whenever(mLabelledProductsSource.invoke(argCaptor.capture())).thenReturn(listOf(labelledProduct1,labelledProduct2,labelledProduct3))
    }

    private fun checkBasicStuff(ra: ResultActions): ResultActions {
        ra.andDo(print())
            .andExpect(status().isOk)
            .andExpect(content().contentType(JSON))

        for(idx in 0..2){
            ra.andExpect(jsonPath("$.products[$idx].productId", `is`("PRODUCT ID ${idx+1}")))
                .andExpect(jsonPath("$.products[$idx].priceLabel", `is`("PRICE LABEL ${idx+1}")))
        }
        return ra;
    }

    private fun doGet(): ResultActions  = mockMvc.perform(get(PATH))

    private fun doGetWithLabelType(labelType: String): ResultActions {
        return mockMvc.perform(get(PATH).param(LABEL_TYPE, labelType))
    }

    @Test
    fun testGetNoLabelTypeQueryString() {
        checkBasicStuff(doGet()).andReturn()
        assertEquals(LabelType.ShowWasNow, argCaptor.firstValue)
    }

    private fun checkGetWithLabelType(type: LabelType) {
        checkBasicStuff(doGetWithLabelType(type.name))
        assertEquals(type, argCaptor.firstValue)
        verify(mLabelledProductsSource, times(1)).invoke(argCaptor.firstValue)
        verifyNoMoreInteractions(mLabelledProductsSource)
    }

    @Test
    fun testWithInvalidLabelType() {
        doGetWithLabelType("BOB").andDo(print())
                .andExpect(status().isUnprocessableEntity)
                .andExpect(content().contentTypeCompatibleWith(MediaType.TEXT_PLAIN))
                .andExpect(content().string("The parameter labelType must have a value among : ShowWasNow, ShowWasThenNow, ShowPercDiscount"))
    }

    @Test
    fun testGetWithWasNow() = checkGetWithLabelType(LabelType.ShowWasNow)

    @Test
    fun testGetWithWasThenNow() = checkGetWithLabelType(LabelType.ShowWasThenNow)

    @Test
    fun testGetWithPercDiscount() = checkGetWithLabelType(LabelType.ShowPercDiscount)
}
